core/node/da_clients/src/avail/sdk.rs

```diff
@@ -369,3 +372,91 @@ fn ss58hash(data: &[u8]) -> Vec<u8> {
     ctx.update(data);
     ctx.finalize().to_vec()
 }
+
+/// An implementation of the `DataAvailabilityClient` trait that interacts with the Avail network.
+#[derive(Debug, Clone)]
+pub(crate) struct GasRelayClient {
+    api_url: String,
+    api_key: String,
+    api_client: Arc<reqwest::Client>,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct GasRelayAPISubmissionResponse {
+    submission_id: String,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct GasRelayAPIStatusResponse {
+    submission: GasRelayAPISubmission,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct GasRelayAPISubmission {
+    block_hash: Option<H256>,
+    extrinsic_index: Option<u64>,
+}
+
+impl GasRelayClient {
+    pub(crate) async fn new(
+        api_url: &str,
+        api_key: &str,
+        api_client: Arc<reqwest::Client>,
+    ) -> anyhow::Result<Self> {
+        Ok(Self {
+            api_url: api_url.to_owned(),
+            api_key: api_key.to_owned(),
+            api_client,
+        })
+    }
+
+    pub(crate) async fn post_data(&self, data: Vec<u8>) -> anyhow::Result<(H256, u64)> {
+        let submit_url = format!("{}/user/submit_raw_data?token=ethereum", &self.api_url);
+        // send the data to the gas relay
+        let submit_response = self
+            .api_client
+            .post(&submit_url)
+            .body(Bytes::from(data))
+            .header("Content-Type", "text/plain")
+            .header("Authorization", &self.api_key)
+            .send()
+            .await
+            .map_err(to_retriable_da_error)?;
+
+        let submit_response = submit_response
+            .json::<GasRelayAPISubmissionResponse>()
+            .await
+            .map_err(to_retriable_da_error)?;
+
+        let status_url = format!(
+            "{}/user/get_submission_info?submission_id={}",
+            self.api_url, submit_response.submission_id
+        );
+
+        let (block_hash, extrinsic_index) = loop {
+            // usually takes around 40s to finalize, but polling every 5s is enough here
+            tokio::time::sleep(tokio::time::Duration::from_secs(5)).await;
```

Comment:
Does it make sense to set max retries/timeout here? If so please use `backon`

Reply:
Current version is bad, please add max retries regardless. If for some reason API doesn't return response it means loop will be stuck and we will have to put some effort into understanding what's going on. If code returns a meaningful error after some failed attempts, it makes it much easier to debug etc

Reply:
Fixed, leaving open if @dimazhornyk wants to implement using `backon`.